package/gluon-geolocator/luasrc/lib/gluon/geolocator/geolocator

```diff
@@ -0,0 +1,130 @@
+#!/usr/bin/lua
+
+local bit = require('bit')
+local unistd = require('posix.unistd')
+local fcntl = require('posix.fcntl')
+local uci = require('simple-uci').cursor()
+local json = require ("jsonc")
+local ubus = require 'ubus'
+local iwinfo = require("iwinfo")
+
+local LOC="gluon-node-info"
+local GLC="geolocator"
+local TIME_STAMP="/tmp/geolocator_timestamp"
+
+if not uci:get_bool(GLC, "settings", "auto_location") then
+	os.exit(0)
+end
+
+-- PID file to ensure the geolocator isn't running parallel
+local lockfile = '/var/lock/geolocator.lock'
+local lockfd, err = fcntl.open(lockfile, bit.bor(fcntl.O_WRONLY, fcntl.O_CREAT), 384) -- mode 0600
+
+if not lockfd then
+	io.stderr:write(err, '\n')
+	os.exit(1)
+end
+
+local ok, _ = fcntl.fcntl(lockfd, fcntl.F_SETLK, {
+	l_start = 0,
+	l_len = 0,
+	l_type = fcntl.F_WRLCK,
+	l_whence = unistd.SEEK_SET,
+})
+
+if not ok then
+	io.stderr:write(string.format(
+		"Unable to lock file %s. Make sure there is no other instance of the geolocator running.\n",
+		lockfile
+	))
+	os.exit(1)
+end
+
+-- Iterates over all active WLAN interfaces
+-- Returning true from the callback function will skip all remaining
+-- interfaces of the same radio
+local function foreach_radio(f)
+	local uconn = assert(ubus.connect(), 'failed to connect to ubus')
+	local status = uconn:call('network.wireless', 'status', {})
+	ubus.close(uconn)
+
+	for _, radio in pairs(status) do
+		for _, iface in ipairs(radio.interfaces) do
+			if f(iface.ifname) then
+				break
+			end
+		end
+	end
+end
+
+local function receive_json(request)
+	local f = assert(io.popen(string.format("exec wget -T 15 -q -O- '%s'", request)), 'failed to run wget')
+	local data = f:read('*a')
+	f:close()
+
+	return json.parse(data)
+end
+
+-- Get position
+local function locate()
+	local done_bssids = {}
+	local found_bssids = {}
+	foreach_radio(function(ifname)
+		local iw = iwinfo[iwinfo.type(ifname)]
+		if not iw then
+			-- Skip other ifaces of this radio, as they
+			-- will have the same type
+			return true
+		end
+
+		local scanlist = iw.scanlist(ifname)
+		if not scanlist then
+			return false
+		end
+
+		for _, entry in ipairs(scanlist) do
+			if entry.mode:match("Master") then
+				local bssid = string.upper(entry.bssid:gsub(":", ""))
+				if not done_bssids[bssid] then
+					table.insert(found_bssids, bssid)
+					done_bssids[bssid] = true
+				end
+			end
+		end
+
+		return true
+	end)
+
+	assert(#found_bssids > 0, 'insufficient BSSIDs found')
+
+	local data = receive_json('http://openwifi.su/api/v1/bssids/' .. table.concat(found_bssids, ','))
```

Comment:
I'm having a hard time with the privacy aspect of announcing your neighbouring WiFi networks in plain over the internet. What's it gonna take to get HTTPS up and running here? Just under 400 kB.

```
root@64289-rocket-carrier:~# opkg info libustream-wolfssl
Package: libustream-wolfssl
Version: 2018-07-30-23a3f283-1
Depends: libc, libubox, libwolfssl
Status: unknown ok not-installed
Section: libs
Architecture: powerpc_8540
Size: 4444
Filename: libustream-wolfssl_2018-07-30-23a3f283-1_powerpc_8540.ipk
Description: ustream SSL Library (wolfssl)

root@64289-rocket-carrier:~# opkg info ca-bundle
Package: ca-bundle
Version: 20180409-2
Depends: libc
Status: unknown ok not-installed
Section: base
Architecture: all
Size: 120447
Filename: ca-bundle_20180409-2_all.ipk
Description: System CA certificates as a bundle

root@64289-rocket-carrier:~# opkg info libwolfssl
Package: libwolfssl
Version: 3.15.3-stable-1
Depends: libc
Provides: libcyassl
Status: install ok installed
Section: libs
Architecture: powerpc_8540
Size: 270414
Filename: libwolfssl_3.15.3-stable-1_powerpc_8540.ipk
Description: wolfSSL (formerly CyaSSL) is an SSL library optimized for small
 footprint, both on disk and for memory use.
Installed-Time: 1549301037
```

And then I noticed the site has a broken HTTPS setup since some time in 2017:

```
openwifi.su uses an invalid security certificate. 
The certificate expired on 23. October 2017, 04:20:00 GMT+2. 
The current time is February 9, 2019, 4:01 AM. Error code: SEC_ERROR_EXPIRED_CERTIFICATE
```

Reply:
I am also not so happy with pushing though data plain over the internet. But the images will be to big with ssl on board. I have an another Idea to only send hashes of SSID+mac but this need time to implement and also time to be filled with datas. So actually there is only this solution.

Reply:
At some point we are going to cross the 4 MB either way. I'm not afraid of that, it just means lower end device will not be able to receive all possible features, which is reasonable, given the size constraint. On the other hand we would be able to build more interesting applications for properly sized routers.

Reply:
just for notice, openwifi.su has fixed the outdated ssl certificate aprox 2019-06-12